File: backend/app/routers/kanban_context.py
"""
Kanban Context API Router

Week 2 Day 5: 3 API endpoints for context management.
Implements Q4 (Double-click auto-load, single-click popup)
"""

from uuid import UUID
from fastapi import APIRouter, Depends, HTTPException, status
from fastapi.responses import JSONResponse

from backend.app.core.security import require_role, UserRole, get_current_user
from backend.app.services.kanban_context_service import kanban_context_service
from backend.app.models.kanban_context import (
    ContextMetadata,
    TaskContext,
    ContextUploadRequest,
    ContextLoadRequest,
    ContextUploadResponse,
    ContextLoadResponse,
    ContextNotFoundError,
    ContextSizeLimitExceeded,
    InvalidContextFiles,
)

router = APIRouter(prefix="/api/kanban/context", tags=["Kanban Context"])


# ============================================================================
# Error Response Helper
# ============================================================================

def error_response(code: str, message: str, status_code: int, details: dict = None):
    """Standard error response format"""
    from datetime import datetime
    return JSONResponse(
        status_code=status_code,
        content={
            "error": {
                "code": code,
                "message": message,
                "details": details or {},
                "timestamp": datetime.utcnow().isoformat() + "Z"
            }
        }
    )


# ============================================================================
# Context Operations (3 endpoints)
# ============================================================================

@router.get(
    "/{task_id}",
    response_model=ContextMetadata,
    dependencies=[Depends(require_role(UserRole.VIEWER))],
    summary="Get context metadata",
    description="Get context metadata without full files list (Q4: Context info)"
)
async def get_context_metadata(
    task_id: UUID,
    current_user: dict = Depends(get_current_user)
):
    """
    Get context metadata for task.

    **RBAC**: Requires `viewer` role or higher.
    **Q4**: Returns metadata for single-click popup.
    **Returns**: ContextMetadata without full files array.
    """
    try:
        metadata = await kanban_context_service.get_context_metadata(task_id)

        if not metadata:
            return error_response(
                code="CONTEXT_NOT_FOUND",
                message=f"No context found for task {task_id}",
                status_code=status.HTTP_404_NOT_FOUND,
                details={"task_id": str(task_id)}
            )

        return metadata

    except Exception as e:
        return error_response(
            code="GET_CONTEXT_FAILED",
            message=f"Failed to get context metadata: {str(e)}",
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR
        )


@router.post(
    "/{task_id}",
    response_model=ContextUploadResponse,
    status_code=status.HTTP_201_CREATED,
    dependencies=[Depends(require_role(UserRole.DEVELOPER))],
    summary="Upload context as ZIP",
    description="Upload context files as ZIP (<50MB limit)"
)
async def upload_context(
    task_id: UUID,
    upload_request: ContextUploadRequest,
    current_user: dict = Depends(get_current_user)
):
    """
    Upload context for task as ZIP file.

    **RBAC**: Requires `developer` role or higher.
    **Q4**: ZIP will be auto-loaded on double-click.
    **Limit**: Max 50MB ZIP size.

    Process:
    1. Validate files list (min 1 file)
    2. Create ZIP from files (mock: estimate size)
    3. Upload to storage (mock: generate URL)
    4. Store metadata in database
    5. Return ZIP URL and checksum
    """
    try:
        upload_response = await kanban_context_service.upload_context(
            task_id, upload_request
        )
        return upload_response

    except ContextSizeLimitExceeded as e:
        return error_response(
            code="CONTEXT_SIZE_LIMIT_EXCEEDED",
            message=str(e),
            status_code=status.HTTP_413_REQUEST_ENTITY_TOO_LARGE,
            details={"task_id": str(task_id), "limit_bytes": 52428800}
        )

    except InvalidContextFiles as e:
        return error_response(
            code="INVALID_CONTEXT_FILES",
            message=str(e),
            status_code=status.HTTP_400_BAD_REQUEST,
            details={"task_id": str(task_id)}
        )

    except Exception as e:
        return error_response(
            code="UPLOAD_CONTEXT_FAILED",
            message=f"Failed to upload context: {str(e)}",
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR
        )


@router.post(
    "/{task_id}/load",
    response_model=ContextLoadResponse,
    dependencies=[Depends(require_role(UserRole.DEVELOPER))],
    summary="Track context load (Q4 double-click)",
    description="Track context load event (Q4: load_count, avg_load_time_ms)"
)
async def track_context_load(
    task_id: UUID,
    load_request: ContextLoadRequest,
    current_user: dict = Depends(get_current_user)
):
    """
    Track context load event (Q4: Double-click auto-load tracking).

    **RBAC**: Requires `developer` role or higher.
    **Q4**: Increments load_count, updates avg_load_time_ms.

    Process:
    1. Find context for task
    2. Increment load_count
    3. Update avg_load_time_ms: (old_avg * old_count + new_time) / new_count
    4. Update last_loaded_at timestamp
    5. Return updated stats
    """
    try:
        load_response = await kanban_context_service.track_context_load(
            task_id, load_request
        )
        return load_response

    except ContextNotFoundError as e:
        return error_response(
            code="CONTEXT_NOT_FOUND",
            message=str(e),
            status_code=status.HTTP_404_NOT_FOUND,
            details={"task_id": str(task_id)}
        )

    except Exception as e:
        return error_response(
            code="TRACK_LOAD_FAILED",
            message=f"Failed to track context load: {str(e)}",
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR
        )


# ============================================================================
# Additional Endpoint: Get Full Context (with files list)
# ============================================================================

@router.get(
    "/{task_id}/full",
    response_model=TaskContext,
    dependencies=[Depends(require_role(UserRole.DEVELOPER))],
    summary="Get full context with files list",
    description="Get complete context including all file paths (developer+ only)"
)
async def get_context_full(
    task_id: UUID,
    current_user: dict = Depends(get_current_user)
):
    """
    Get full context including files array.

    **RBAC**: Requires `developer` role or higher.
    **Returns**: Full TaskContext with files array.
    **Use case**: Download context for local development.
    """
    try:
        context = await kanban_context_service.get_context_full(task_id)

        if not context:
            return error_response(
                code="CONTEXT_NOT_FOUND",
                message=f"No context found for task {task_id}",
                status_code=status.HTTP_404_NOT_FOUND,
                details={"task_id": str(task_id)}
            )

        return context

    except Exception as e:
        return error_response(
            code="GET_FULL_CONTEXT_FAILED",
            message=f"Failed to get full context: {str(e)}",
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR
        )

File: web-dashboard/lib/api/kanban-context.ts
/**
 * Kanban Context API Client
 *
 * Week 2 Day 4: Context operations (ZIP upload/download, Q4 implementation)
 *
 * API Endpoints:
 * - GET  /api/kanban/context/{id}       - Get metadata
 * - POST /api/kanban/context/{id}       - Upload ZIP
 * - POST /api/kanban/context/{id}/load  - Track load event
 * - GET  /api/kanban/context/{id}/full  - Get full context
 */

const API_BASE_URL = process.env.NEXT_PUBLIC_API_URL || 'http://localhost:8000'

export interface ContextMetadata {
  task_id: string
  file_count: number
  total_size_bytes: number
  zip_url?: string
  checksum?: string
  load_count: number
  avg_load_time_ms: number
  created_at: string
  updated_at: string
  last_loaded_at?: string
}

export interface ContextFile {
  path: string
  size_bytes: number
  mime_type?: string
}

export interface TaskContext extends ContextMetadata {
  files: ContextFile[]
}

export interface ContextUploadRequest {
  files: ContextFile[]
  checksum?: string
}

export interface ContextUploadResponse {
  task_id: string
  zip_url: string
  checksum: string
  file_count: number
  total_size_bytes: number
  created_at: string
}

export interface ContextLoadRequest {
  load_time_ms: number
}

export interface ContextLoadResponse {
  task_id: string
  load_count: number
  avg_load_time_ms: number
  last_loaded_at: string
}

export class KanbanContextAPIError extends Error {
  constructor(
    message: string,
    public statusCode?: number,
    public code?: string
  ) {
    super(message)
    this.name = 'KanbanContextAPIError'
  }
}

/**
 * Get context metadata (single-click popup)
 */
export async function fetchContextMetadata(
  taskId: string
): Promise<ContextMetadata | null> {
  const response = await fetch(
    `${API_BASE_URL}/api/kanban/context/${taskId}`,
    {
      method: 'GET',
      headers: {
        'Content-Type': 'application/json',
      },
    }
  )

  if (response.status === 404) {
    return null // No context found
  }

  if (!response.ok) {
    const errorData = await response.json().catch(() => ({}))
    throw new KanbanContextAPIError(
      errorData.error?.message || `Failed to fetch context metadata: ${response.statusText}`,
      response.status,
      errorData.error?.code
    )
  }

  return response.json()
}

/**
 * Upload context as ZIP (<50MB limit)
 */
export async function uploadContext(
  taskId: string,
  uploadRequest: ContextUploadRequest
): Promise<ContextUploadResponse> {
  const response = await fetch(
    `${API_BASE_URL}/api/kanban/context/${taskId}`,
    {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
      },
      body: JSON.stringify(uploadRequest),
    }
  )

  if (!response.ok) {
    const errorData = await response.json().catch(() => ({}))

    // Special handling for size limit exceeded
    if (response.status === 413) {
      throw new KanbanContextAPIError(
        'Context size limit exceeded (50MB max)',
        413,
        'CONTEXT_SIZE_LIMIT_EXCEEDED'
      )
    }

    throw new KanbanContextAPIError(
      errorData.error?.message || `Failed to upload context: ${response.statusText}`,
      response.status,
      errorData.error?.code
    )
  }

  return response.json()
}

/**
 * Upload context ZIP file (Week 6 Day 5: FormData file upload)
 * @param taskId - Task ID to upload context for
 * @param file - ZIP file to upload
 * @returns Upload response with metadata
 * @throws KanbanContextAPIError if upload fails or file exceeds 50MB
 */
export async function uploadContextFile(
  taskId: string,
  file: File
): Promise<ContextUploadResponse> {
  // Validate file size (50MB = 52,428,800 bytes)
  const MAX_SIZE = 50 * 1024 * 1024
  if (file.size > MAX_SIZE) {
    throw new KanbanContextAPIError(
      `File size ${(file.size / 1024 / 1024).toFixed(2)}MB exceeds 50MB limit`,
      413,
      'CONTEXT_SIZE_LIMIT_EXCEEDED'
    )
  }

  // Validate file type (ZIP only)
  if (!file.name.endsWith('.zip') && file.type !== 'application/zip') {
    throw new KanbanContextAPIError(
      'Only ZIP files are allowed',
      400,
      'INVALID_FILE_TYPE'
    )
  }

  const formData = new FormData()
  formData.append('file', file)

  const response = await fetch(
    `${API_BASE_URL}/api/kanban/context/${taskId}/upload`,
    {
      method: 'POST',
      body: formData,
      // Note: Don't set Content-Type header - browser will set it with boundary
    }
  )

  if (!response.ok) {
    const errorData = await response.json().catch(() => ({}))

    // Special handling for size limit exceeded
    if (response.status === 413) {
      throw new KanbanContextAPIError(
        'Context size limit exceeded (50MB max)',
        413,
        'CONTEXT_SIZE_LIMIT_EXCEEDED'
      )
    }

    throw new KanbanContextAPIError(
      errorData.error?.message || `Failed to upload context: ${response.statusText}`,
      response.status,
      errorData.error?.code
    )
  }

  return response.json()
}

/**
 * Track context load event (Q4: double-click auto-load)
 */
export async function trackContextLoad(
  taskId: string,
  loadRequest: ContextLoadRequest
): Promise<ContextLoadResponse> {
  const response = await fetch(
    `${API_BASE_URL}/api/kanban/context/${taskId}/load`,
    {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
      },
      body: JSON.stringify(loadRequest),
    }
  )

  if (!response.ok) {
    const errorData = await response.json().catch(() => ({}))
    throw new KanbanContextAPIError(
      errorData.error?.message || `Failed to track context load: ${response.statusText}`,
      response.status,
      errorData.error?.code
    )
  }

  return response.json()
}

/**
 * Get full context with files list (for download)
 */
export async function fetchFullContext(
  taskId: string
): Promise<TaskContext | null> {
  const response = await fetch(
    `${API_BASE_URL}/api/kanban/context/${taskId}/full`,
    {
      method: 'GET',
      headers: {
        'Content-Type': 'application/json',
      },
    }
  )

  if (response.status === 404) {
    return null // No context found
  }

  if (!response.ok) {
    const errorData = await response.json().catch(() => ({}))
    throw new KanbanContextAPIError(
      errorData.error?.message || `Failed to fetch full context: ${response.statusText}`,
      response.status,
      errorData.error?.code
    )
  }

  return response.json()
}

/**
 * Download context ZIP file (browser download)
 */
export async function downloadContextZip(
  taskId: string,
  zipUrl: string,
  filename?: string
): Promise<void> {
  try {
    const response = await fetch(zipUrl)

    if (!response.ok) {
      throw new KanbanContextAPIError(
        `Failed to download context ZIP: ${response.statusText}`,
        response.status
      )
    }

    const blob = await response.blob()
    const url = window.URL.createObjectURL(blob)
    const link = document.createElement('a')
    link.href = url
    link.download = filename || `context-${taskId}.zip`
    document.body.appendChild(link)
    link.click()
    document.body.removeChild(link)
    window.URL.revokeObjectURL(url)
  } catch (error) {
    if (error instanceof KanbanContextAPIError) {
      throw error
    }
    throw new KanbanContextAPIError(
      `Failed to download context ZIP: ${error instanceof Error ? error.message : 'Unknown error'}`
    )
  }
}
